lua/nvim-treesitter/fold.lua

```diff
@@ -1,28 +1,70 @@
+local api = vim.api
+local utils = require'nvim-treesitter.ts_utils'
+local query = require'nvim-treesitter.query'
 local parsers = require'nvim-treesitter.parsers'
 
 local M = {}
 
-function M.get_fold_indic(lnum)
-  if not parsers.has_parser() or not lnum then return '0' end
+-- This is cached on buf tick to avoid computing that multiple times
+-- Especially not for every line in the file when `zx` is hit
+local folds_levels = utils.memoize_by_buf_tick(function(bufnr)
+  local lang = parsers.get_buf_lang(bufnr)
+
+  local matches
+  if query.has_fold(lang) then
+    matches = query.get_capture_matches(bufnr, "@fold", "fold")
+  elseif query.has_locals(lang) then
+    matches = query.get_capture_matches(bufnr, "@scope", "locals")
+  else
+    return {}
+  end
+
+  local levels_tmp = {}
 
-  local function smallest_multiline_containing(node, level)
-    for index = 0,(node:named_child_count() -1) do
-      local child = node:named_child(index)
-      local start, _, stop, _ = child:range()
+  for _, node in ipairs(matches) do
+    local start, _, stop, stop_col = node.node:range()
 
-      if start ~= stop and start <= (lnum -1) and stop >= (lnum -1) then
-        return smallest_multiline_containing(child, level + 1)
-      end
+    if stop_col > 0 then
+      stop = stop + 1
```

Comment:
A this marks also an exclusive end of the fold... The stop col *might* not be necessary.

Reply:
Maybe ... We'll try later on